pmlc/conversion/gpu/gpu_to_vulkan.cc

```diff
@@ -237,10 +263,115 @@ Value ConvertGpuLaunchFuncToVulkanCalls::createEntryPointNameConstant(
                                   shaderName, LLVM::Linkage::Internal);
 }
 
+llvm::SmallSet<Operation *, 4>
+ConvertGpuLaunchFuncToVulkanCalls::getExternalDependentOperations(Value value) {
+  llvm::SmallSet<Operation *, 4> operations;
+  llvm::SetVector<OpOperand *> uses;
+
+  operations.insert(value.getDefiningOp());
+  for (auto &use : value.getUses()) {
+    uses.insert(&use);
+  }
+  while (!uses.empty()) {
+    auto owner = uses.back()->getOwner();
+    operations.insert(owner);
+    uses.pop_back();
+    for (auto opResult : owner->getOpResults()) {
+      for (auto &use : opResult.getUses()) {
+        uses.insert(&use);
+      }
+    }
+  }
+
+  for (auto operation : operations) {
+    if (isInternalOperation(operation)) {
+      operations.erase(operation);
+    }
+  }
+  return operations;
+}
+
+bool ConvertGpuLaunchFuncToVulkanCalls::isInternalOperation(Operation *op) {
+  // Buffer related calls are all mlir::CallOp, does not need to check
+  // LLVM::CallOp
+  auto callOp = llvm::dyn_cast<mlir::CallOp>(op);
+  if (callOp && (callOp.callee() == kBindAllBuffers ||
+                 optionalSymbols.count(callOp.callee()))) {
+    return true;
+  }
+
+  auto allocOp = llvm::dyn_cast<AllocOp>(op);
+  auto memRefCastOp = llvm::dyn_cast<MemRefCastOp>(op);
+  if (allocOp || memRefCastOp) {
+    return true;
+  }
+  return false;
+}
+
+BufferCopyMode
+ConvertGpuLaunchFuncToVulkanCalls::getBufferCopyMode(mlir::CallOp callOp,
+                                                     Value buffer) {
+  BufferCopyMode copyMode = BufferCopyMode::NoCopy;
+
+  if (buffer.isa<BlockArgument>()) {
+    copyMode = copyMode |
+               (BufferCopyMode::HostToDevice | BufferCopyMode::DeviceToHost);
+    return copyMode;
+  }
+
+  auto currentBlock = callOp.getOperation()->getBlock();
+
+  // Get llvm.call @run in current block
+  LLVM::CallOp *pRunOp = nullptr;
```

Comment:
Don't create pointers to `CallOp`, instead use it by-value.

Like:
```
LLVM::CallOp runOp = nullptr;
```

And later like:
```
runOp = llvmCallOp;
```

Reply:
Even better to extract this to a new function like:

```
Optional<LLVM::CallOp> findCallOp(llvm::StringRef name) {
}
```

Reply:
Fixed, added a findLLVMCallOp, since here only checks LLVM::CallOp